src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt

```diff
@@ -172,38 +174,48 @@ class ADRController(private val encryption: Encryption,
                       @PathVariable resourceType: String,
                       @PathVariable modelCalibrateId: String,
                       @RequestParam resourceFileName: String,
+                      @RequestParam resourceName: String,
                       @RequestParam resourceId: String?): ResponseEntity<String>
     {
-        // 1. Download relevant artefact from hintr
-        val artefact = when (resourceType)
+        // 1. If output file, download relevant artefact from hintr
+        val artefact: Pair<ResponseEntity<StreamingResponseBody>, String>? = when (resourceType)
         {
             appProperties.adrOutputZipSchema -> Pair(apiClient.downloadSpectrum(modelCalibrateId),
                     "Naomi model outputs")
             appProperties.adrOutputSummarySchema -> Pair(apiClient.downloadSummary(modelCalibrateId),
                     "Naomi summary report")
+            appProperties.adrPJNZSchema, appProperties.adrShapeSchema, appProperties.adrPopSchema,
+            appProperties.adrSurveySchema, appProperties.adrARTSchema, appProperties.adrANCSchema -> null
             else -> return ErrorDetail(HttpStatus.BAD_REQUEST, "Invalid resourceType").toResponseEntity()
         }
 
-        // 2. Return error if artefact can't be retrieved
-        if (!artefact.first.statusCode.is2xxSuccessful)
+        // 2. Get the file ready to upload
+        val tmpDir = Files.createTempDirectory("adr").toFile()
+        val (file, errorDetail) = if (artefact != null)
         {
-            val baos = ByteArrayOutputStream()
-            artefact.first.body?.writeTo(baos)
-            return ErrorDetail(artefact.first.statusCode, baos.toString()).toResponseEntity()
+            makeOutputFileToPushToADR(artefact, tmpDir, resourceFileName)
+        }
+        else
+        {
+            makeInputFileToPushToADR(resourceType, resourceId, tmpDir)
         }
 
-        // 3. Stream artefact to file
-        val tmpDir = Files.createTempDirectory("adr").toFile()
-        val file = File(tmpDir, resourceFileName)
-        FileOutputStream(file).use { fis ->
-            artefact.first.body!!.writeTo(fis)
+        if (errorDetail != null)
+        {
+           return errorDetail.toResponseEntity()
         }
 
-        // 4. Checksum file and upload with metadata to ADR
-        val filePart = Pair("upload", file)
+        // 3. Checksum file and upload with metadata to ADR
+        @Suppress("UnsafeCallOnNullableType")
```

Comment:
I think this @suppress("UnsafeCallOnNullableType") can be avoided.

Reply:
How would you avoid it? Need to do the unsafe call, either here or when we call md5sum - we know from the logic of the called method that exactly one of errorDetail and file will be null, but the compiler doesn't pick that up.